Energy computed instead of reading

jotakar opened the pull request:
Some modules reads energy in kwh with precission of 1 digit  (2,1 kwh, 2,2 kwh ...) but sometimes the energy consumed in a day is less than 0.1 kwh. 
Think about a lamp type led or low consum: 8 watts for 10 hours is 80 wh (0,080kwh). but module reports 0.0 kwh for this day. 
With this modification system gets the energy value by software: it reads the usage and the interval of time and make a simple operation:  usage*time (usage: watts, time: minutes). 
I use field Options for Energy Meter devices. If  enabled (1)  the energy is calculated by program if  disabled (0) energy value is read from the device. 
I've found this issue with Qubino relays.

jotakar committed:
Energy computed instead of read

Some modules read energy in kwh with precission of 1 digit  (2,1 kwh, 2,2 kwh ...) but sometimes the energy consumed in a day is less than 0.1 kwh. Think about a lamp type led or low consum: 8 watts for 10 hours is 80 wh (0,080kwh). but module reports 0.0 kwh for this day. 
With this modification system gets the energy value by software: it reads the usage and the interval of time and make a simple operation: usage*time (usage: watts, time: minutes). 
I use field Options for Energy Meter devices. If  enabled (1)  the energy is calculated by program if  disabled (0) energy value is read from device. 
I've found this issue with Qubino relays.
Other files affected: www/views/utility.html, www/app/UtiliController.js

jotakar committed:
Energy computed instead of read 

Some modules read energy in kwh with precission of 1 digit  (2,1 kwh, 2,2 kwh ...) but sometimes the energy consumed in a day is less than 0.1 kwh. 

Think about a lamp type led or low consum: 8 watts for 10 hours is 80 wh (0,080kwh). but module reports 0.0 kwh for this day. 
With this modification system gets the energy value by software: it reads the usage and the interval of time and make a simple operation: 

usage*time (usage: watts, time: minutes). 
I use field Options for Energy Meter devices. If  enabled (1)  the energy is calculated by program if  disabled (0) energy value is read from 

device. 
I've found this issue with Qubino relays.
Other files affected main/SQLHelper.cpp  and  www/app/UtiliController.js

jotakar committed:
Energy computed instead of read

Some modules read energy in kwh with precission of 1 digit  (2,1 kwh, 2,2 kwh ...) but sometimes the energy consumed in a day is less than 0.1 kwh. 

Think about a lamp type led or low consum: 8 watts for 10 hours is 80 wh (0,080kwh). but module reports 0.0 kwh for this day. 
With this modification system gets the energy value by software: it reads the usage and the interval of time and make a simple operation: 

usage*time (usage: watts, time: minutes). 
I use field Options for Energy Meter devices. If  enabled (1)  the energy is calculated by program if  disabled (0) energy value is read from 
device. 
I've found this issue with Qubino relays.
Other files affected: main/SQLhelper.cpp and www/views/utility.html

jotakar committed:
Energy computed instead of read

Some words translated into spanish.

gizmocuz commented:
Thank you for the patch, but as you might have noticed both our build systems have problems with compile errors.

gizmocuz commented:
You can not change the language files, please provide here (as text) the needed changes to the language file

jotakar committed:
SQLhelper fixed

sOption is now declared
strptime doesn't exist for windows, so I've convert time string to time_t  straight.

jotakar commented:
The words added and their translation into spanish are:
Energy read: Lectura de energía
Computed: Calculada
From device: Desde el módulo

Thanks

gizmocuz commented:
Why did you remove the 'HTTPPostContentType' in SQLHelper.cpp ?

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -3095,8 +3090,9 @@ unsigned long long CSQLHelper::UpdateValueInt(const int HardwareID, const char*
 	std::vector<std::vector<std::string> > result;
-	result = safe_query("SELECT ID,Name, Used, SwitchType, nValue, sValue FROM DeviceStatus WHERE (HardwareID=%d AND DeviceID='%q' AND Unit=%d AND Type=%d AND SubType=%d)",HardwareID, ID, unit, devType, subType);
-	if (result.size()==0)
+	result = safe_query("SELECT ID,Name, Used, SwitchType, nValue, sValue, LastUpdate, Description, Options FROM DeviceStatus WHERE (HardwareID=%d AND DeviceID='%q' AND Unit=%d AND Type=%d AND SubType=%d)",HardwareID, ID, unit, devType, subType);
```
Do we need to query for Description?

jotakar replied:
No, I had to have deleted Description. I'll delete it.

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -3455,7 +3473,7 @@ unsigned long long CSQLHelper::UpdateValueInt(const int HardwareID, const char*
 		break;
 	}
-	if (bDeviceUsed)
+if (bDeviceUsed)
```
missing a tab here... (that was previously here)

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -4017,8 +4035,11 @@ void CSQLHelper::UpdateRainLog()
+			int rate=0;
+
+			rate=atoi(splitresults[0].c_str());
+			total = static_cast<float>(atof(splitresults[1].c_str()));
```
revert these changes please, they are not needed and only give cppcheck issues
This patch is about energy/power anyway ;)

jotakar replied:
Sorry, I uploaded my version of SQLHelper.cpp and it had this code, so I reverted some changes made after my versión (5231). I'll revert all code I did'nt modified. You're right: this patch is just for energy/power.
Sorry it's my first pull.

gizmocuz replied:
No problem... ;)

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -4126,10 +4147,7 @@ void CSQLHelper::UpdateUVLog()
-		pTypeUV,
-		pTypeGeneral, sTypeUV
-	);
+	result=safe_query("SELECT ID,Type,SubType,nValue,sValue,LastUpdate FROM DeviceStatus WHERE (Type=%d)", pTypeUV);
```
Revert this ? Why did you do this ?

jotakar replied:
Like the upper comment.

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -4169,7 +4187,7 @@ void CSQLHelper::UpdateUVLog()
 			safe_query(
 				"INSERT INTO UV (DeviceRowID, Level) "
-				"VALUES ('%llu', '%g')",
+				"VALUES ('%llu', '%.1f')",
```
revert this, why did you do this ?

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -5388,7 +5406,7 @@ void CSQLHelper::AddCalendarUpdateUV()
 			result=safe_query(
 				"INSERT INTO UV_Calendar (DeviceRowID, Level, Date) "
-				"VALUES ('%llu', '%g', '%q')",
+				"VALUES ('%llu', '%.2f', '%q')",
```
Revert this, why did you do this ?

jotakar committed:
Revert accidental changes

I hope this patch is good. If it doesn't ... begin again?, from cero?

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -5388,7 +5411,7 @@ void CSQLHelper::AddCalendarUpdateUV()
 			result=safe_query(
 				"INSERT INTO UV_Calendar (DeviceRowID, Level, Date) "
-				"VALUES ('%llu', '%g', '%q')",
+				"VALUES ('%llu', '%.g', '%q')",
```
I still see a lot of non-needed changes, and for example this one (procent point 'g') is not what it should be
Maybe best to revert this whole file, and only make that change needed for your patch
So we also do not get extra lines, or extra spaces at the end of the lines

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -3132,17 +3133,40 @@ unsigned long long CSQLHelper::UpdateValueInt(const int HardwareID, const char*
+	            StringSplit(sValue, ";", parts);
+	            sprintf(sCompValue, "%s;%.0f", parts[0].c_str(),nEnergy);
+	            sValue = sCompValue;
+            }
```
How do you know it is an Energy device here ? All devices can have sOption="1"

jotakar replied:
Well I can use devType and subType , that's what I've done in last modif.

jotakar committed:
Delete SQLHelper.cpp

jotakar committed:
Create SQLHelper.cpp

jotakar committed:
Rewirte clean SQLHelper.cpp

I've made the changes in a clean copy of SQLHelper.cpp from domoticz/domoticz master. 
The only changes mus be in lines 3098, 3135, 3145-3166.

jotakar committed:
Webserver cpp Energy computed

WEbServer must return field sOptions for alternate energy meter (energy computed or straight)

jotakar commented:
Hi
Which is the status of this pull request?

jotakar commented:
Hi
I'm using this modification in my net and it's working fine. 
Is it being tested by anyone else?

YcKe commented:
Seems like this would solve indeed the Qubino metering precision issue. Great work jotakar! Hope it gets merged soon.

jotakar commented:
Thanks
I suppose it's being tested or so. I'm using it and it works fine in my net.

gizmocuz commented:
Thanks

jotakar commented:
I am pleased to participate in this program.
